components/merkledb/src/views/metadata.rs

```diff
@@ -212,135 +261,128 @@ impl<T: IndexAccess> IndexesPool<T> {
         self.0.get(&()).unwrap_or_default()
     }
 
-    fn set_len(&mut self, len: u64) {
-        self.0.put(&(), len)
+    fn index_metadata(&self, index_name: &[u8]) -> Option<IndexMetadata> {
+        self.0.get(index_name)
     }
 
-    fn index_metadata<V>(&self, index_name: &[u8]) -> Option<IndexMetadata<V>>
-    where
-        V: BinaryAttribute + Default + Copy,
-    {
-        self.0.get(index_name)
+    fn set_len(&mut self, len: u64) {
+        self.0.put_or_forget(&(), len);
     }
 
     fn create_index_metadata<V>(
         &mut self,
         index_name: &[u8],
         index_type: IndexType,
-    ) -> IndexMetadata<V>
+    ) -> (IndexMetadata<V>, bool)
```

Comment:
Nit: I think it deserves an enum: `Phantom`/`Real` is more readable than `true`/`false` (especially when you realize that `true` means phantom index).